### PR TITLE
Add random batch rnn

### PR DESCRIPTION
**Pull request by jinbeizame007**

### CODE REVIEW

File: machina/traj/traj.py [hunk 1]
```diff
@@ -278,6 +278,64 @@ def random_batch(self, batch_size, epoch=1, indices=None, return_indices=False):
...
+            keys = seqs[0].keys()
+            for key in keys:
+                batch[key] = torch.cat([seq[key].unsqueeze(0)
+                                        for seq in seqs], dim=0)
+                # (batch_size, seq_length, *) -> (seq_length, batch_size, *)
+                batch[key] = batch[key].transpose(0, 1)
+            out_masks = torch.ones(
+                (seq_length, batch_size), dtype=torch.float, device=get_device())
+            for i in range(batch_size):
+                out_masks[lengths[i]:-1, i] = 0
```
Comment (rarilurelo) [inline machina/traj/traj.py]: The last value of this sequence should be masked, so this line may be replaced by `out_masks[lengths[i]:, i] = 0`.

File: machina/traj/traj.py [hunk 2]
```diff
@@ -278,6 +278,64 @@ def random_batch(self, batch_size, epoch=1, indices=None, return_indices=False):
...
+                0, len(self._epis_index)-1, (batch_size,))
+
+            for idx in indices:
+                length = min(
+                    self._epis_index[idx+1] - self._epis_index[idx], seq_length)
+                start = np.random.randint(
+                    self._epis_index[idx], self._epis_index[idx+1] - length + 1)
+                data_map = dict()
+                for key in self.data_map:
+                    if idx == len(self._epis_index)-2 and length != seq_length:
```
Comment (rarilurelo) [inline machina/traj/traj.py]: if seq_length is larger than the sum of the lengths of last two episodes, indexing of Line 319 will be failed. I think `if self._epis_index[-1] - self._epis_index[idx] < seq_length:` may be correct solution.

File: machina/traj/traj.py [hunk 3]
```diff
@@ -278,6 +278,64 @@ def random_batch(self, batch_size, epoch=1, indices=None, return_indices=False):
...
+    def random_batch_rnn(self, batch_size, seq_length=None, epoch=1):
+        """
+        Providing sequences of batch which is randomly sampled from trajectory.
+        batch shape is (seq_length, batch_size, *)
+
+        Parameters
+        ----------
+        batch_size : int
+        seq_length : int
+            Length of sequence of batch.
```
Comment (rarilurelo) [inline machina/traj/traj.py]: Could you add comment for `None`?

File: machina/traj/traj.py [hunk 4]
```diff
@@ -278,6 +278,65 @@ def random_batch(self, batch_size, epoch=1, indices=None, return_indices=False):
...
+
+            for idx in indices:
+                length = min(
+                    self._epis_index[idx+1] - self._epis_index[idx], seq_length)
+                start = np.random.randint(
+                    self._epis_index[idx], self._epis_index[idx+1] - length + 1)
+                data_map = dict()
+                for key in self.data_map:
+                    if self._epis_index[-1] - self._epis_index[idx] < seq_length:
+                        pad = torch.zeros_like(self.data_map[key], dtype=torch.float, device=get_device())[
```
Comment (rarilurelo) [inline machina/traj/traj.py]: I think this zero array generation is memory inefficient. `torch.zeros_like(self.data_map[key][:seq_length - length]` may be more efficient.

**Comment by rarilurelo** [comment]
LGTM!
Thank you for adding iteration method to Traj.
